integration-test/src/python/local_test_runner/test_template.py

```diff
@@ -52,56 +54,54 @@ def __init__(self, testname, params):
   # pylint: disable=too-many-return-statements, too-many-branches,
   # pylint: disable=too-many-statements
   def run_test(self):
-    """ Runs the test template """
-
+    """ Runs the test template. Must either return TestSuccess or raise TestFailure"""
+    topology_submitted = False
     try:
       # prepare test data, start the topology and block until it's running
       self._prepare_test_data()
       self.submit_topology()
+      topology_submitted = True
       _block_until_stmgr_running(self.get_expected_container_count())
 
-      if not self._block_until_topology_running(self.get_expected_min_instance_count()):
-        self.cleanup_test()
-        return False
+      self._block_until_topology_running(self.get_expected_min_instance_count())
 
       # Execute the specific test logic and block until topology is running again
       self.execute_test_case()
 
       _block_until_stmgr_running(self.get_expected_container_count())
       physical_plan_json =\
         self._block_until_topology_running(self.get_expected_min_instance_count())
-      if not physical_plan_json:
-        self.cleanup_test()
-        return False
 
       # trigger the test data to flow and invoke the pre_check_results hook
       self._inject_test_data()
-      if not self.pre_check_results(physical_plan_json):
-        self.cleanup_test()
-        return False
-    except Exception as e:
-      logging.error("Test failed, attempting to clean up: %s", e)
+      self.pre_check_results(physical_plan_json)
+
+      # finally verify the expected results
+      result = self._check_results()
       self.cleanup_test()
-      return False
+      return result
 
-    # finally verify the expected results
-    return self._check_results()
+    except status.TestFailure as e:
+      if topology_submitted:
+        logging.error("Test failed, attempting to clean up")
+        self.cleanup_test()
+      raise e
+    except Exception as e:
```

Comment:
That's ok since `e` is scoped within each except clause so only 1 will be defined without collision.